cache/contenthash/checksum.go

```diff
@@ -515,8 +535,25 @@ func (cc *cacheContext) wildcards(ctx context.Context, m *mount, p string) ([]*W
 			updated = true
 		}
 
-		wildcards = append(wildcards, &Wildcard{Path: string(fn), Record: cr})
+		if cr.Type == CacheRecordTypeDir {
+			lastIncludedDir = fn
+
+			if excludePatternMatcher != nil {
+				dirSlash := fn + "/"
+				for _, pat := range excludePatternMatcher.Patterns() {
+					patStr := pat.String() + "/"
+					if strings.HasPrefix(patStr, dirSlash) {
```

Comment:
As curently implemented this is best-effort, and if we have a pattern here rather than a prefix we might end up including some paths in the checksum that would be excluded by the filters. This would never cause false cache hits but might lead to false negatives.

The alternative would be to never use a dir's content hash when there is at least one exclude pattern, and instead walk all included dirs and only incorporate non-excluded files in the final checksum. Basically this code would be replaced with:

```go
if cr.Type == CacheRecordTypeDir {
	lastIncludedDir = fn

	if excludePatternMatcher != nil {
		// don't use the optimization where we add a whole dir's contents
		// to the checksum when there is an exclude pattern, because some
		// files in the dir may be excluded
		continue treeWalk
	}
}
```